File: 2025/11월_5주차/BOJ17413_단어_뒤집기_2.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    string S;
    // 공백 포함해서 한 줄 전체 입력
    getline(cin, S);

    string word;   // 태그 밖에서 단어를 임시로 저장
    string result; // 최종 결과를 쌓을 문자열
    bool inTag = false; // 지금 태그 안에 있는지 여부

    for (char c : S) {
        if (c == '<') {
            // 태그 시작 전에 버퍼에 쌓인 단어가 있으면 뒤집어서 추가
            if (!word.empty()) {
                reverse(word.begin(), word.end());
                result += word;
                word.clear();
            }
            inTag = true;
            result += c; // '<' 까지 그대로 추가
        }
        else if (c == '>') {
            inTag = false;
            result += c; // '>' 도 그대로 추가
        }
        else if (inTag) {
            // 태그 안에서는 어떤 문자든 그대로 추가
            result += c;
        }
        else {
            // 태그 밖의 문자 처리
            if (c == ' ') {
                // 단어 종료 → 뒤집어서 추가
                if (!word.empty()) {
                    reverse(word.begin(), word.end());
                    result += word;
                    word.clear();
                }
                result += c; // 공백 그대로
            } else {
                // 단어 구성 중인 문자
                word += c;
            }
        }
    }

    // 문자열이 끝났을 때, 마지막 단어가 남아 있을 수 있음
    if (!word.empty()) {
        reverse(word.begin(), word.end());
        result += word;
    }

    cout << result << '\n';
    return 0;
}

File: 2025/11월_5주차/BOJ17413_단어_뒤집기_2.py
import sys
S = sys.stdin.readline().rstrip()

in_tag = False
word = []
result = []

for c in S:
    if c == '<':
        # 태그 시작 전에 word가 비어있지 않으면 뒤집어서 출력
        if word:
            result.append(''.join(reversed(word)))
            word = []
        in_tag = True
        result.append(c)

    elif c == '>':
        in_tag = False
        result.append(c)

    elif in_tag:
        # 태그 안은 그대로 출력
        result.append(c)

    else:
        # 태그 밖
        if c == ' ':
            # 공백에서 단어 종료 → 뒤집어서 출력
            if word:
                result.append(''.join(reversed(word)))
                word = []
            result.append(' ')   # 공백 그대로 출력
        else:
            # 단어 구성 중
            word.append(c)

# 마지막 단어 처리
if word:
    result.append(''.join(reversed(word)))

print(''.join(result))
